fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/PlanOrderingKey.java

```diff
@@ -283,10 +286,18 @@ private static KeyExpression orderingCompatiblePlanKey(@Nonnull PlanOrderingKey
                 if (pos < 0) {
                     return null;
                 }
-                if (planOrderingKey.isEqualityBound(pos)) {
-                    // Equality bound predicate. We're trivially ordered by it, so add it to the final key.
-                    // When combining sub-plans, this can represent, slotting in all of the elements of one
-                    // sub-plan before or after the elements of another sub-plan.
+                if (pos < nextNonPrefix) {
+                    // Two possibilities:
+                    // 1. This key component appears in the equality bound prefix. If that's the
+                    //    case, then we're trivially ordered by it
+                    // 2. This key component appears in the non-equality bound tail, and we've already
+                    //    matched this key.
+                    // In either case, the ordering key is already satisfied, so we can add it to the
+                    // final component list.
+                    //
+                    // Note that we can get an already bound equality predicate if when combining sub-plans,
+                    // we need to slot in all the results from one sub-plan before or after the elements of
+                    // another sub-plan in order to satisfy the plans overall ordering requirement.
                     components.add(component);
```

Comment:
This is the key, and I'm not sure if I've explained it correctly. The basic conceit here is that if the key component is less than `nextNonPrefix`, then either:

1. The key component is equality bound, and so we're the same as before. All results in the plan will have the same value for the expression, so we're "ordered" by it trivially.
2. The key component is not equality bound, but we're already ordered by it. That means that results for the plan is ordered in a way such that the plan and the candidate key are ordered in a way means that they are the same up to some prefix which must include this key component. So, we only need to validate that the _suffixes_ are compatibly ordered for a fixed value of the prefix (up to `nextNonPrefix`). For such a grouping, all results have the same value for this expression, so we are likewise correctly ordered.

I'm not sure on how to best explain this, though. Hopefully, that tracks, because it would be nice to validate that this fix is conceptually correct